arelle/plugin/validate/NL/PluginValidationDataExtension.py

```diff
@@ -21,3 +24,34 @@ class PluginValidationDataExtension(PluginValidationData):
     documentResubmissionUnsurmountableInaccuraciesQn: QName
     entrypointRoot: str
     entrypoints: set[str]
+
+    _contextsByDocument: dict[str, list[ModelContext]] | None = None
+    _factsByDocument: dict[str, list[ModelFact]] | None = None
+    _unitsByDocument: dict[str, list[ModelUnit]] | None = None
+
+    def contextsByDocument(self, modelXbrl: ModelXbrl) -> dict[str, list[ModelContext]]:
+        if self._contextsByDocument is not None:
+            return self._contextsByDocument
+        contextsByDocument = defaultdict(list)
+        for context in modelXbrl.contexts.values():
+            contextsByDocument[context.modelDocument.filepath].append(context)
+        self._contextsByDocument = dict(contextsByDocument)
```

Comment:
What does this line do? Isnt contextsByDocument already a dict?

Reply:
It's a `defaultdict`. I didn't want subsequent calls to this function to write to the dict.